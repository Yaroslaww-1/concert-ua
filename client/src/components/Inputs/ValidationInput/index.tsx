import React from 'react';

import styles from './styles.module.scss';
import { TextField } from '@material-ui/core';

export interface IValidationInputProps {
  id: string;
  label: string;
  type?: 'search' | 'password';
  onEdit?: (isValid: boolean, newInput: string) => void;
  defaultValue?: string;
  validateInput?: (newInput: string) => Error | null;
  classes?: {
    root?: string;
    textField?: string;
  };
}

const ValidationInput: React.FC<IValidationInputProps> = ({
  id,
  label,
  type = 'text',
  onEdit = () => {},
  validateInput = () => null,
  defaultValue = '',
  classes = { root: '', textField: '' },
}) => {
  const [inputValue, setInputValue] = React.useState<string>('');
  const [focused, setFocused] = React.useState<boolean>(false);
  const [error, setError] = React.useState<Error | null>(null);

  React.useEffect(() => {
    setInputValue(defaultValue);
  }, [defaultValue]);

  const onChange = (event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
    const { value } = event.target;
    setInputValue(value);
    const error = validateInput(value);
    setError(error);
    onEdit(error === null, value);
  };

  return (
    <div className={`${styles.root} ${classes.root}`}>
      <TextField
        id={id}
        placeholder={label}
        type={type}
        variant="outlined"
        classes={{ root: `${styles.textFieldRoot} ${classes.textField}` }}
        value={inputValue}
        onChange={onChange}
        error={error && !focused ? true : false}
        helperText={error && !focused ? error.message : ''}
        FormHelperTextProps={{ className: styles.helperText }}
        onBlur={() => setFocused(false)}
        onFocus={() => setFocused(true)}
      />
    </div>
  );
};

export default ValidationInput;
