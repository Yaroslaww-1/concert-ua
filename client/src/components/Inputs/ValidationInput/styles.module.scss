@import 'src/styles/variables';

.root {
  .textFieldRoot {
    display: flex;
    flex-direction: column;
    width: 100%;
    height: 100%;
    * {
      color: $colorBlack;
    }
    * fieldset {
      border: none;
    }
    * [aria-invalid=true] {
      border: 1px solid $colorRed;
    }
    * [aria-invalid=false] {
      border: 1px solid $colorMiddleLightGray;
    }
    .helperText {
      position: absolute;
      background-color: $colorRed;
      top: -25px;
      left: -14px;
      padding: 0px 10px 0px 10px;
      height: 40%;
      color: $colorWhite;
      text-align: center;
      vertical-align: middle;
      font-size: 0.85rem;
    }
  }
}