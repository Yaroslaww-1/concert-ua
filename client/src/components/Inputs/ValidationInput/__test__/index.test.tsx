import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import ValidationInput from '../index';

describe('ValidationInput', () => {
  const setup = ({
    onEdit = () => {},
    validateInput = () => null,
  }: {
    onEdit?: (isValid: boolean) => void;
    validateInput?: () => Error | null;
  }) => {
    const { container } = render(
      <ValidationInput id="id" label="label" onEdit={onEdit} validateInput={validateInput} />,
    );
    const input = container.getElementsByTagName('input').item(0)!;
    return {
      input,
      ...container,
    };
  };

  it('should contain input', () => {
    const { input } = setup({});
    const value = 'custom input value';
    fireEvent.change(input, { target: { value } });
    expect(input.value).toBe(value);
  });

  it('should display error on error input', () => {
    const onEdit = jest.fn((isValid) => isValid);
    const error = new Error('Custom error');
    const { input } = setup({ onEdit, validateInput: () => error });

    const value = 'custom input value';
    fireEvent.change(input, { target: { value } });

    const mockResults = onEdit.mock.results;
    expect(mockResults[mockResults.length - 1].value).toBe(false);
  });
});
