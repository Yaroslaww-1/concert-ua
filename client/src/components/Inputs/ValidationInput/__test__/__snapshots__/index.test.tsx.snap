// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextComponent should match snapshot 1`] = `
<div
  class="root "
>
  <div
    class="MuiFormControl-root MuiTextField-root textFieldRoot "
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input"
        id="id"
        placeholder="label"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
        style="padding-left: 8px;"
      >
        <legend
          class="PrivateNotchedOutline-legend-2"
          style="width: 0.01px;"
        >
          <span>
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
    
  </div>
</div>
`;
